File: mbw/src/main/java/com/mycelium/wallet/PinDialog.java
/*
 * Copyright 2013, 2014 Megion Research and Development GmbH
 *
 * Licensed under the Microsoft Reference Source License (MS-RSL)
 *
 * This license governs use of the accompanying software. If you use the software, you accept this license.
 * If you do not accept the license, do not use the software.
 *
 * 1. Definitions
 * The terms "reproduce," "reproduction," and "distribution" have the same meaning here as under U.S. copyright law.
 * "You" means the licensee of the software.
 * "Your company" means the company you worked for when you downloaded the software.
 * "Reference use" means use of the software within your company as a reference, in read only form, for the sole purposes
 * of debugging your products, maintaining your products, or enhancing the interoperability of your products with the
 * software, and specifically excludes the right to distribute the software outside of your company.
 * "Licensed patents" means any Licensor patent claims which read directly on the software as distributed by the Licensor
 * under this license.
 *
 * 2. Grant of Rights
 * (A) Copyright Grant- Subject to the terms of this license, the Licensor grants you a non-transferable, non-exclusive,
 * worldwide, royalty-free copyright license to reproduce the software for reference use.
 * (B) Patent Grant- Subject to the terms of this license, the Licensor grants you a non-transferable, non-exclusive,
 * worldwide, royalty-free patent license under licensed patents for reference use.
 *
 * 3. Limitations
 * (A) No Trademark License- This license does not grant you any rights to use the Licensor’s name, logo, or trademarks.
 * (B) If you begin patent litigation against the Licensor over patents that you think may apply to the software
 * (including a cross-claim or counterclaim in a lawsuit), your license to the software ends automatically.
 * (C) The software is licensed "as-is." You bear the risk of using it. The Licensor gives no express warranties,
 * guarantees or conditions. You may have additional consumer rights under your local laws which this license cannot
 * change. To the extent permitted under your local laws, the Licensor excludes the implied warranties of merchantability,
 * fitness for a particular purpose and non-infringement.
 */

package com.mycelium.wallet;

import android.content.Context;
import android.os.Build;
import android.os.Handler;
import android.os.Message;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatDialog;

import com.google.common.base.Strings;
import com.mycelium.wallet.activity.settings.helper.TwoFactorHelper;
import com.mycelium.wallet.activity.util.FingerprintHandler;
import com.mycelium.wallet.activity.util.Pin;

import java.util.ArrayList;
import java.util.Collections;

import kotlin.Unit;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function1;

public class PinDialog extends AppCompatDialog {
   public static final String PLACEHOLDER_TYPED = "\u25CF"; // Unicode Character 'BLACK CIRCLE' (which is a white circle in our dark theme)
   public static final String PLACEHOLDER_NOT_TYPED = "\u25CB"; // Unicode Character 'WHITE CIRCLE' (which is a black circle)
   public static final String PLACEHOLDER_SMALL = "\u2022"; // Unicode Character  'BULLET'
   protected Button btnBack;
   protected Button btnClear;
   private FingerprintHandler fingerprintHandler;
   private TwoFactorHelper twoFactorHelper;

   public interface OnPinEntered {
      void pinEntered(PinDialog dialog, Pin pin);
   }

   public interface FingerprintCallback {
      void onSuccess();
   }

   protected ArrayList<Button> buttons = new ArrayList<Button>(10);
   protected ArrayList<TextView> disps = new ArrayList<TextView>(6);
   protected String enteredPin;

   protected OnPinEntered onPinValid = null;
   protected FingerprintCallback fingerprintCallback;
   private boolean hidden;
   protected boolean pinPadIsRandomized;
   protected boolean isTwoFactorAuth;

   public void setOnPinValid(OnPinEntered _onPinValid) {
      this.onPinValid = _onPinValid;
      twoFactorHelper.setListener(_onPinValid);
   }

   public void setFingerprintCallback(FingerprintCallback fingerprintCallback) {
      this.fingerprintCallback = fingerprintCallback;
   }

   public PinDialog(Context context, boolean hidden, boolean cancelable) {
      super(context);
      getWindow().setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE);
      pinPadIsRandomized = MbwManager.getInstance(context).isPinPadRandomized();
      isTwoFactorAuth = MbwManager.getInstance(context).isTwoFactorEnabled();
      this.hidden = hidden;
      setCancelable(cancelable);
      setCanceledOnTouchOutside(false);
      loadLayout();
      initPinPad();
      enteredPin = "";
      clearDigits();
      updatePinDisplay();
      this.setTitle(R.string.pin_enter_pin);
      initFingerprint(context);
      initTwoFactorAuthentication();
   }

   private void initTwoFactorAuthentication() {

   }

   private void initFingerprint(Context context) {
      final View view = findViewById(R.id.fingerprintHint);
      TextView logicView  = findViewById(R.id.logicOperationHint);

      if (view != null) {
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P
                 && MbwManager.getInstance(context).isFingerprintEnabled()
                 && FingerprintHandler.Companion.isFingerprintAvailable(context)) {
            fingerprintHandler = new FingerprintHandler();
            fingerprintHandler.startAuth(context, new Function0<Unit>() {
               @Override
               public Unit invoke() {
                  if (isTwoFactorAuth) {
                     view.setEnabled(false);
                     twoFactorHelper.fingerprintSuccess();
                  } else {
                     dismiss();
                     if (fingerprintCallback != null) {
                        fingerprintCallback.onSuccess();
                     }
                  }
                  return null;
               }
            }, new Function1<String, Unit>() {
               @Override
               public Unit invoke(String msg) {
                  Toast.makeText(getContext(), msg, Toast.LENGTH_SHORT);
                  return null;
               }
            });
            logicView.setText(isTwoFactorAuth ? R.string.and : R.string.or);
         } else {
            view.setVisibility(View.GONE);
            logicView.setVisibility(View.GONE);
         }
      }
      twoFactorHelper = new TwoFactorHelper(this);
   }

   @Override
   public void dismiss() {
      super.dismiss();
      if (fingerprintHandler != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
         fingerprintHandler.cancelAuth();
      }
   }

   protected void initPinPad() {
      disps.add((TextView) findViewById(R.id.pin_char_1));
      disps.add((TextView) findViewById(R.id.pin_char_2));
      disps.add((TextView) findViewById(R.id.pin_char_3));
      disps.add((TextView) findViewById(R.id.pin_char_4));
      disps.add((TextView) findViewById(R.id.pin_char_5));
      disps.add((TextView) findViewById(R.id.pin_char_6));
      buttons.add( ((Button) findViewById(R.id.pin_button0)));
      buttons.add( ((Button) findViewById(R.id.pin_button1)));
      buttons.add( ((Button) findViewById(R.id.pin_button2)));
      buttons.add( ((Button) findViewById(R.id.pin_button3)));
      buttons.add( ((Button) findViewById(R.id.pin_button4)));
      buttons.add( ((Button) findViewById(R.id.pin_button5)));
      buttons.add( ((Button) findViewById(R.id.pin_button6)));
      buttons.add( ((Button) findViewById(R.id.pin_button7)));
      buttons.add( ((Button) findViewById(R.id.pin_button8)));
      buttons.add( ((Button) findViewById(R.id.pin_button9)));

      ArrayList<Integer> numbers = new ArrayList<>();
      for (int i = 0; i < 10; i++) {
         numbers.add(i);
      }
      if (pinPadIsRandomized) {
         Collections.shuffle(numbers);
      }
      for (int i = 0; i < 10; i++) {
         buttons.get(i).setText(numbers.get(i).toString());
      }

      btnClear = findViewById(R.id.pin_clr);
      btnBack = findViewById(R.id.pin_back);

      for (Button b : buttons) {
         final int num = Integer.parseInt(b.getText().toString());
         b.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
               addDigit(String.valueOf(num));
            }
         });
      }

      btnBack.setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View v) {
            removeLastDigit();
         }
      });

      btnClear.setOnClickListener(new View.OnClickListener() {
         @Override
         public void onClick(View v) {
            clearDigits();
            updatePinDisplay();
         }
      });
   }

   protected void loadLayout() {
      setContentView(R.layout.enter_pin_dialog);
   }

   protected void addDigit(String c) {
      enteredPin = enteredPin + c;
      updatePinDisplay();
   }

   protected void updatePinDisplay(){
      int cnt = 0;
      for (TextView t : disps) {
         t.setText(getPinDigitAsString(enteredPin, cnt));
         cnt++;
      }
      checkPin();
   }

   protected String getPinDigitAsString(String pin, int index) {
      if (pin.length() > index) {
         return hidden ? PLACEHOLDER_TYPED : pin.substring(index, index + 1);
      } else {
         return hidden ? PLACEHOLDER_NOT_TYPED : PLACEHOLDER_SMALL;
      }
   }

   protected void clearDigits() {
      enteredPin = "";
      for (TextView t : disps) {
         t.setText(hidden ? PLACEHOLDER_NOT_TYPED : PLACEHOLDER_SMALL);
      }
   }

   protected void removeLastDigit(){
      if (!Strings.isNullOrEmpty(enteredPin)){
         enteredPin = enteredPin.substring(0, enteredPin.length() - 1);
      }
      updatePinDisplay();
   }

   protected void enableButtons(boolean enabled) {
      for (Button b : buttons) {
         b.setEnabled(enabled);
      }
   }

   protected void checkPin() {
      if (enteredPin.length() >= 6) {
         acceptPin();
      }
   }

   protected void acceptPin() {
      enableButtons(false);
      delayHandler.sendMessage(delayHandler.obtainMessage());
   }

   /**
    * Trick to make the last digit update before the dialog is disabled
    */
   final Handler delayHandler = new Handler() {
      public void handleMessage(Message msg) {
         if (isTwoFactorAuth) {
            btnClear.setEnabled(false);
            btnBack.setEnabled(false);
            twoFactorHelper.pinEntered(getPin());
         } else {
            if (onPinValid != null) onPinValid.pinEntered(PinDialog.this, getPin());
            enableButtons(true);
            clearDigits();
         }
      }
   };

   protected Pin getPin() {
      return new Pin(enteredPin);
   }
}

File: mbw/src/main/java/com/mycelium/wallet/PinDialog.kt
 package com.mycelium.wallet

import android.content.Context
import android.os.Handler
import android.os.Message
import android.view.WindowManager
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatDialog
import androidx.core.view.isVisible
import androidx.fragment.app.FragmentActivity
import com.google.common.base.Strings
import com.mycelium.wallet.activity.modern.Toaster
import com.mycelium.wallet.activity.settings.helper.TwoFactorHelper
import com.mycelium.wallet.activity.util.FingerprintHandler
import com.mycelium.wallet.activity.util.Pin
import com.mycelium.wallet.databinding.EnterPinDialogBinding
import com.mycelium.wallet.databinding.EnterPinDisplayBinding
import com.mycelium.wallet.databinding.EnterPinNumpadBinding

open class PinDialog(context: Context, val hidden: Boolean, cancelable: Boolean) :
    AppCompatDialog(context) {
    private val fingerprintHandler = FingerprintHandler().apply {
        successListener = {
            if (isTwoFactorAuth) {
//                  fingerprintHint.setEnabled(false);
                numpadBinding?.pinFinger?.isVisible = false
                twoFactorHelper.fingerprintSuccess()
            } else {
                dismiss()
                if (fingerprintCallback != null) {
                    fingerprintCallback!!.onSuccess()
                }
            }
        }
        failListener = { msg: String? ->
            Toaster(context).toast(msg!!, false)
        }
    }
    private val twoFactorHelper = TwoFactorHelper(this)
    protected var numpadBinding: EnterPinNumpadBinding? = null
    protected var pinBinding: EnterPinDisplayBinding? = null

    interface OnPinEntered {
        fun pinEntered(dialog: PinDialog, pin: Pin)
    }

    interface FingerprintCallback {
        fun onSuccess()
    }

    var buttons = listOf<Button>()
    protected var disps = listOf<TextView>()
    protected var enteredPin: String = ""
    var pinValidCallback: OnPinEntered? = null
    var fingerprintCallback: FingerprintCallback? = null
    protected var pinPadIsRandomized = MbwManager.getInstance(context).isPinPadRandomized
    protected var isTwoFactorAuth = MbwManager.getInstance(context).isTwoFactorEnabled
    protected var isFingerprintEnabled = MbwManager.getInstance(context).isFingerprintEnabled

    open fun setOnPinValid(_onPinValid: OnPinEntered?) {
        pinValidCallback = _onPinValid
        twoFactorHelper.listener = _onPinValid
    }


    protected fun initFingerprint(context: Context) {
        if (isFingerprintEnabled) {
            val result = fingerprintHandler.authenticate(context)
            if (!result) {
                Toaster(getContext()).toast(R.string.fingerprint_not_available, false)
                numpadBinding?.pinFinger?.isVisible = false
            }
        }
        numpadBinding?.pinFinger?.isVisible = enteredPin.isEmpty() && isFingerprintEnabled
    }

    override fun onStart() {
        super.onStart()

    }

    override fun onStop() {
        fingerprintHandler.successListener = null
        fingerprintHandler.failListener = null
        super.onStop()
    }

    protected open fun initPinPad() {
        disps = pinBinding?.let {
            listOf(it.pinChar1, it.pinChar2, it.pinChar3, it.pinChar4, it.pinChar5, it.pinChar6)
        }.orEmpty()
        buttons = numpadBinding?.let {
            listOf(
                it.pinButton0, it.pinButton1, it.pinButton2, it.pinButton3, it.pinButton4,
                it.pinButton5, it.pinButton6, it.pinButton7, it.pinButton8, it.pinButton9
            )
        }.orEmpty()
        (0..9).toList().let {
            if (pinPadIsRandomized) it.shuffled()
            else it
        }.forEachIndexed { i, value ->
            buttons[i].text = value.toString()
        }
        buttons.forEach {
            val num = it.text.toString().toInt()
            it.setOnClickListener { addDigit(num.toString()) }
        }
        numpadBinding?.pinBack?.setOnClickListener {
            removeLastDigit()
        }
        numpadBinding?.pinClr?.setOnClickListener {
            clearDigits()
            updatePinDisplay()
        }
    }

    protected open fun loadLayout() {
        setContentView(EnterPinDialogBinding.inflate(layoutInflater).apply {
            numpadBinding = this.keyboard.numPad
            pinBinding = this.keyboard.pinDisplay
        }.root)
        twoFactorHelper.needFingerCallback = {
            initFingerprint(context)
            numpadBinding?.pinFinger?.isVisible = true
        }
        numpadBinding?.pinFinger?.setOnClickListener {
            initFingerprint(context)
        }
    }

    protected fun addDigit(c: String) {
        enteredPin += c
        updatePinDisplay()
    }

    protected open fun updatePinDisplay() {
        disps.forEachIndexed { index, textView ->
            textView.text = getPinDigitAsString(enteredPin, index)
        }
        checkPin()
        numpadBinding?.pinFinger?.isVisible = enteredPin.isEmpty() && isFingerprintEnabled
    }

    protected fun getPinDigitAsString(pin: String, index: Int): String =
        if (pin.length > index) {
            if (hidden) PLACEHOLDER_TYPED else pin.substring(index, index + 1)
        } else {
            if (hidden) PLACEHOLDER_NOT_TYPED else PLACEHOLDER_SMALL
        }

    protected fun clearDigits() {
        enteredPin = ""
        disps.forEach {
            it.text = if (hidden) PLACEHOLDER_NOT_TYPED else PLACEHOLDER_SMALL
        }
    }

    protected fun removeLastDigit() {
        if (!Strings.isNullOrEmpty(enteredPin)) {
            enteredPin = enteredPin.substring(0, enteredPin.length - 1)
        }
        updatePinDisplay()
    }

    protected fun enableButtons(enabled: Boolean) {
        buttons.forEach {
            it.isEnabled = enabled
        }
    }

    protected open fun checkPin() {
        if (enteredPin.length >= 6) {
            acceptPin()
        }
    }

    protected fun acceptPin() {
        enableButtons(false)
        delayHandler.sendMessage(delayHandler.obtainMessage())
    }

    /**
     * Trick to make the last digit update before the dialog is disabled
     */
    val delayHandler: Handler = object : Handler() {
        override fun handleMessage(msg: Message) {
            if (isTwoFactorAuth) {
                numpadBinding?.pinClr?.isEnabled = false
                numpadBinding?.pinBack?.isEnabled = false
                twoFactorHelper.pinEntered(Pin(enteredPin))
            } else {
                // Optimized to not cause 4 bytes difference in DEX thus failing build reproducibility check
                if (pinValidCallback != null) {
                    val _pinDialog = this@PinDialog
                    pinValidCallback?.pinEntered(
                        _pinDialog,
                        Pin(enteredPin)
                    )
                }
                enableButtons(true)
                clearDigits()
            }
        }
    }

    init {
        window!!.setFlags(
            WindowManager.LayoutParams.FLAG_SECURE,
            WindowManager.LayoutParams.FLAG_SECURE
        )
        setCancelable(cancelable)
        setCanceledOnTouchOutside(false)
        loadLayout()
        initPinPad()
        clearDigits()
        updatePinDisplay()
        this.setTitle(R.string.pin_enter_pin)

        fingerprintHandler.onCreate(context as FragmentActivity)

        initFingerprint(context)
    }

    companion object {
        const val PLACEHOLDER_TYPED =
            "\u25CF" // Unicode Character 'BLACK CIRCLE' (which is a white circle in our dark theme)
        const val PLACEHOLDER_NOT_TYPED =
            "\u25CB" // Unicode Character 'WHITE CIRCLE' (which is a black circle)
        const val PLACEHOLDER_SMALL = "\u2022" // Unicode Character  'BULLET'
    }
}